src/supernova/mod.rs

```diff
@@ -0,0 +1,1202 @@
+    let (nifs_primary, (r_U_primary_folded, r_W_primary_folded)) =
+      if let Some((r_U_primary, r_W_primary)) = self
+        .r_U_primary
+        .get(claim.get_augmented_circuit_index())
+        .unwrap_or(&None)
+        .as_ref()
+        .zip(
+          self
+            .r_W_primary
+            .get(claim.get_augmented_circuit_index())
+            .unwrap_or(&None)
+            .as_ref(),
+        )
+      {
+        NIFS::prove(
+          ck_primary,
+          &pp.ro_consts_primary,
+          &self.pp_digest,
+          &pp.r1cs_shape_primary,
+          r_U_primary,
+          r_W_primary,
+          &l_u_primary,
+          &l_w_primary,
+        )?
+      } else {
+        NIFS::prove(
+          ck_primary,
+          &pp.ro_consts_primary,
+          &self.pp_digest,
+          &pp.r1cs_shape_primary,
+          &RelaxedR1CSInstance::default(ck_primary, &pp.r1cs_shape_primary),
+          &RelaxedR1CSWitness::default(&pp.r1cs_shape_primary),
+          &l_u_primary,
+          &l_w_primary,
+        )?
+      };
+
+    let mut cs_secondary: SatisfyingAssignment<G2> = SatisfyingAssignment::new();
+    let binding = Commitment::<G1>::decompress(&nifs_primary.comm_T)?;
+    let inputs_secondary: SuperNovaAugmentedCircuitInputs<'_, G1> =
+      SuperNovaAugmentedCircuitInputs::new(
+        self.pp_digest,
+        G2::Scalar::from(self.i as u64),
+        z0_secondary,
+        Some(&self.zi_secondary),
+        Some(&self.r_U_primary),
+        Some(&l_u_primary),
+        Some(&binding),
+        G2::Scalar::ZERO, // secondary circuit never constrain/bump program counter
+        G2::Scalar::from(claim.get_augmented_circuit_index() as u64),
+      );
+
+    let circuit_secondary: SuperNovaAugmentedCircuit<'_, G1, C2> = SuperNovaAugmentedCircuit::new(
+      &pp.augmented_circuit_params_secondary,
+      Some(inputs_secondary),
+      c_secondary,
+      pp.ro_consts_circuit_secondary.clone(),
+      self.num_augmented_circuits,
+    );
+    let (_, zi_secondary) = circuit_secondary
+      .synthesize(&mut cs_secondary)
+      .map_err(|_| NovaError::SynthesisError)?;
+
+    let (l_u_secondary_next, l_w_secondary_next) = cs_secondary
+      .r1cs_instance_and_witness(&pp.r1cs_shape_secondary, ck_secondary)
+      .map_err(|_| NovaError::UnSat)?;
+
+    // update the running instances and witnesses
+    let zi_primary = zi_primary
+      .iter()
+      .map(|v| v.get_value().ok_or(NovaError::SynthesisError))
+      .collect::<Result<Vec<<G1 as Group>::Scalar>, NovaError>>()?;
+    let zi_primary_pc_next = zi_primary_pc_next
+      .get_value()
+      .ok_or(NovaError::SynthesisError)?;
+    let zi_secondary = zi_secondary
+      .iter()
+      .map(|v| v.get_value().ok_or(NovaError::SynthesisError))
+      .collect::<Result<Vec<<G2 as Group>::Scalar>, NovaError>>()?;
+
+    if zi_primary.len() != pp.F_arity_primary || zi_secondary.len() != pp.F_arity_secondary {
+      return Err(NovaError::InvalidStepOutputLength);
+    }
```

Comment:
Same idea: this check is probably better placed as early as we can.

Reply:
ditto and fixed in https://github.com/microsoft/Nova/pull/204/commits/61dad5b531b7316313b991d9651a32584286a6c6